### PR TITLE
Support posting (not replies) to channels and one on one chat

### PR DESCRIPTION
**Pull request by solomongifford**
Support posting (not replies) to channels and one on one chat

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,4 +22,4 @@ Optional:
 1. `HUBOT_OFFICE365_TENANT_FILTER` - Comma seperated list of Office365 tenant Ids that are allowed to communicate with your hubot. By default ALL Office365 tenants can communicate with your hubot if they sideload your application manifest.
 
 # Contributing
-This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comm
\ No newline at end of file
+This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comm
```
Comment (MattSFT) [inline README.md]: Could you undo this change.

File: src/adapter.coffee [hunk 2]
```diff
@@ -50,10 +51,22 @@ class BotFrameworkAdapter extends Adapter
...
+        #    payload = @using(context).toSendable(context, msg)
+        #    if !Array.isArray(payload)
+        #        payload = [payload]
+        #    @connector.send payload, (err, _) ->
+        #        if err
+        #            throw err
+
     reply: (context, messages...) ->
         @robot.logger.info "#{LogPrefix} reply"
+        console.log context
```
Comment (MattSFT) [inline src/adapter.coffee]: remove

File: src/msteams-middleware.coffee [hunk 3]
```diff
@@ -47,6 +47,15 @@ class MicrosoftTeamsMiddleware extends BaseMiddleware
             return null
 
         address = activity.address
+        #for v,k of activity
```
Comment (MattSFT) [inline src/msteams-middleware.coffee]: remove

File: src/msteams-middleware.coffee [hunk 4]
```diff
@@ -63,14 +72,21 @@ class MicrosoftTeamsMiddleware extends BaseMiddleware
                 getMentions(activity),
                 activity.address.bot.id,
                 @robot.name)
+            if -1 is message.text.indexOf @robot.name
```
Comment (MattSFT) [inline src/msteams-middleware.coffee]: use startsWith instead of indexOf. Makes this more legible

File: src/msteams-middleware.coffee [hunk 5]
```diff
@@ -63,14 +72,21 @@ class MicrosoftTeamsMiddleware extends BaseMiddleware
...
 
     toSendable: (context, message) ->
         @robot.logger.info "#{LogPrefix} toSendable"
         msg = message
-        conversationAddress = context.user.activity.address
+        conversationAddress = ''
+        if context.user.activity.address?
+             conversationAddress = context.user.activity.address
+        else
+             conversationAddress = context.user.directroom
```
Comment (MattSFT) [inline src/msteams-middleware.coffee]: Does this work? The address is a larger object than just an Id.

File: src/adapter.coffee [hunk 6]
```diff
@@ -50,10 +51,22 @@ class BotFrameworkAdapter extends Adapter
 
     send: (context, messages...) ->
         @robot.logger.info "#{LogPrefix} send"
+        if context.room? and -1 isnt context.room.indexOf "http"
```
Comment (MattSFT) [inline src/adapter.coffee]: startsWith instead of indexOf

File: src/adapter.coffee [hunk 7]
```diff
@@ -50,10 +51,22 @@ class BotFrameworkAdapter extends Adapter
...
+            @messageRoom context, messages...
+            return
         @reply context, messages...
 
+        #for msg in messages
```
Comment (MattSFT) [inline src/adapter.coffee]: remove the commented code

**Comment by solomongifford** [comment]
The BotFramework does not allow sending a message to a room except via webhook.  Hubot's code for messageRoom:

`
 messageRoom: (room, strings...) ->
    envelope = { room: room }
    @adapter.send envelope, strings...
`

So assuming that "room" is a webhook url, the adapter fails because context.user.activity.address is undefined. Thus the changes to support sending to a room (via webhook).

**Comment by maxpert** [comment]
So if you look at code closely you will see following send loop:

```coffeescript
        for msg in messages
            activity = context.user.activity
            payload = @using(activity.source).toSendable(context, msg)
            if !Array.isArray(payload)
                payload = [payload]
            @connector.send payload, (err, _) ->
                if err
                    throw err
```

The `@using(activity.source).toSendable(context, msg)` here converts a messages from your bot to a sendable message for Directline, as that is the payload (that btw can have your room info) forwarded to Directline as is (it can be array of messages sure). So if there is a specific kind of message that talks to room, it should be in middleware (feel free to contribute to middle-wares) and not the adapter itself. Your change-set breaks whole bunch of stuff for everyone else.
